README.md
# IndexedDB KeyStore

[![NPM](https://img.shields.io/npm/v/keystore-idb)](https://www.npmjs.com/package/keystore-idb)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/fission-suite/blob/master/LICENSE)
[![Maintainability](https://api.codeclimate.com/v1/badges/b0fabd7e80c6bd2c0c7b/maintainability)](https://codeclimate.com/github/fission-suite/keystore-idb/maintainability)
[![Built by FISSION](https://img.shields.io/badge/⌘-Built_by_FISSION-purple.svg)](https://fission.codes)
[![Discord](https://img.shields.io/discord/478735028319158273.svg)](https://discord.gg/zAQBDEq)
[![Discourse](https://img.shields.io/discourse/https/talk.fission.codes/topics)](https://talk.fission.codes)

In-browser key management with IndexedDB and the Web Crypto API.

Securely store and use keys for encryption, decryption, and signatures. IndexedDB and Web Crypto keep keys safe from malicious javascript.

Supports both RSA (RSASSA-PKCS1-v1_5 & RSA-OAEP) and Elliptic Curves (P-256, P-381 & P-521).

ECC (Elliptic Curve Cryptography) is only available on Chrome. Firefox and Safari do not support ECC and must use RSA.
_Specifically, this is an issue with storing ECC keys in IndexedDB_



## Config

Below is the default config and all possible values
_Note: these are given as primitives, but in Typescript you can use the included enums_

```typescript
const defaultConfig = {
  type: 'ecc', // 'ecc' | 'rsa'
  curve: 'P-256', // 'P-256' | 'P-384' | 'P-521'
  rsaSize: 2048, // 1024 | 2048 | 4096
  symmAlg: 'AES-CTR', // 'AES-CTR' | 'AES-GCM' | 'AES-CBC'
  symmLen: 128, // 128 | 192 | 256
  hashAlg: 'SHA-256', // 'SHA-1' | 'SHA-256' | 'SHA-384' | 'SHA-512'
  charSize: 16, // 8 | 16
  storeName: 'keystore', // any string
  exchangeKeyName: 'exchange-key', // any string
  writeKeyName: 'write-key', // any string
}
```
_Note: if you don't include a crypto "type" (`'ecc' | 'rsa'`), the library will check if your browser supports ECC. If so (Chrome), it will use ECC, if not (Firefox, Safari) it will fall back to RSA._



## Example Usage

```typescript
import keystore from 'keystore-idb'

async function run() {
  await keystore.clear()

  const ks1 = await keystore.init({ storeName: 'keystore' })
  const ks2 = await keystore.init({ storeName: 'keystore2' })

  const msg = "Incididunt id ullamco et do."

  // exchange keys and write keys are separate because of the Web Crypto API
  const exchangeKey1 = await ks1.publicExchangeKey()
  const writeKey1 = await ks1.publicWriteKey()
  const exchangeKey2 = await ks2.publicExchangeKey()

  // these keys get exported as strings
  console.log('exchangeKey1: ', exchangeKey1)
  console.log('writeKey1: ', writeKey1)
  console.log('exchangeKey2: ', exchangeKey2)

  const sig = await ks1.sign(msg)
  const valid = await ks2.verify(msg, sig, writeKey1)
  console.log('sig: ', sig)
  console.log('valid: ', valid)

  const cipher = await ks1.encrypt(msg, exchangeKey2)
  const decipher = await ks2.decrypt(cipher, exchangeKey1)
  console.log('cipher: ', cipher)
  console.log('decipher: ', decipher)
}

run()
```



## Development

```shell
# install dependencies
yarn

# run development server
yarn start

# build
yarn build

# test
yarn test

# test w/ reloading
yarn test:watch

# publish (run this script instead of npm publish!)
./publish.sh
```


CHANGELOG.md
# Changelog

### v0.15.5

- Support a `max` parameter (based on rejection sampling) in `randomBuf` rng.

### v0.15.4

- Bump one-webcrypto to 1.0.3 for wider bundler support.

### v0.15.3

- Internally use the one-webcrypto library for referring to the webcrypto API

### v0.15.2

- Add `AES-GCM` to the list of valid symmetric algorithms (`SymmAlg`)
- Internally dynamically refer to either the NodeJS or Browser webcrypto API

### v0.15.1

Importing `keystore-idb/lib/*` directly should now work as intended. This allows bundlers to use the "real" import paths (eg. `import "keystore-idb/lib/utils.js"`) in addition to the "proxy" import paths (eg. `import "keystore-idb/utils.js"`). One reason to do this could be that you want your library to support both new and old bundlers, ie. bundlers with or without `exports` support in their `package.json` file.


### v0.15.0

- Renamed read key to exchange key.
- Switched out `Buffer` usage with `uint8arrays` library.
- Built with esbuild instead of rollup.



### v0.14.0

Use the `globalThis` global object instead of `window`.


src/types.ts
export type Msg = Uint8Array | string | ArrayBuffer

export type CipherText = Uint8Array
export type SymmKey = CryptoKey

export type PublicKey = CryptoKey
export type PrivateKey = CryptoKey

export type Config = {
  type: CryptoSystem
  curve: EccCurve
  rsaSize: RsaSize
  symmAlg: SymmAlg
  symmLen: SymmKeyLength
  hashAlg: HashAlg
  storeName: string
  exchangeKeyName: string
  writeKeyName: string
}

export type SymmKeyOpts = {
  alg: SymmAlg
  length: SymmKeyLength
  iv: Uint8Array
}

export enum CryptoSystem {
  ECC = 'ecc',
  RSA = 'rsa',
}

export enum EccCurve {
  P_256 = 'P-256',
  P_384 = 'P-384',
  P_521 = 'P-521',
}

export enum RsaSize {
  B1024 = 1024,
  B2048 = 2048,
  B4096 = 4096
}

export enum SymmAlg {
  AES_CTR = 'AES-CTR',
  AES_CBC = 'AES-CBC',
  AES_GCM = 'AES-GCM',
}

export enum SymmKeyLength {
  B128 = 128,
  B192 = 192,
  B256 = 256,
}

export enum HashAlg {
  SHA_1 = 'SHA-1',
  SHA_256 = 'SHA-256',
  SHA_384 = 'SHA-384',
  SHA_512 = 'SHA-512',
}

export enum KeyUse {
  Exchange = 'exchange',
  Write = 'write',
}

export interface KeyStore {
  cfg: Config

  exchangeKey: () => Promise<CryptoKeyPair>
  writeKey: () => Promise<CryptoKeyPair>
  getSymmKey: (keyName: string, cfg?: Partial<Config>) => Promise<CryptoKey>
  keyExists(keyName: string): Promise<boolean>
  deleteKey(keyName: string): Promise<void>
  destroy(): Promise<void>

  // Symmetric

  importSymmKey(
    keyStr: string,
    keyName: string,
    cfg?: Partial<Config>
  ): Promise<void>

  exportSymmKey(
    keyName: string,
    cfg?: Partial<Config>
  ): Promise<string>

  encryptWithSymmKey(
    msg: string,
    keyName: string,
    cfg?: Partial<Config>
  ): Promise<string>

  decryptWithSymmKey(
    cipherBytes: string,
    keyName: string,
    cfg?: Partial<Config>
  ): Promise<string>

  // Asymmetric

  sign(
    msg: string,
    cfg?: Partial<Config>
  ): Promise<string>

  verify(
    msg: string,
    sig: string,
    publicKey: string,
    cfg?: Partial<Config>
  ): Promise<boolean>

  encrypt(
    msg: string,
    publicKey: string,
    cfg?: Partial<Config>
  ): Promise<string>

  decrypt(
    cipherText: string,
    publicKey: string,
    cfg?: Partial<Config>
  ): Promise<string>

  publicExchangeKey(): Promise<string>
  publicWriteKey(): Promise<string>
}
